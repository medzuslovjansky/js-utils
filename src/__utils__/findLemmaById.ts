import fs from 'node:fs';
import path from 'node:path';

const fixturesDir = path.join(__dirname, '../../src/__fixtures__');

let fixtures: Map<string, string[]> | undefined;

export function findLemmaById(id: string): string {
  return initFixtures().get(id)?.[2] ?? id;
}

function initFixtures() {
  if (!fixtures) {
    const map = (fixtures = new Map());
    fs.readdirSync(fixturesDir).forEach((name) => {
      const jsonPath = path.join(fixturesDir, name);
      const json = fs.readFileSync(jsonPath, 'utf-8');
      const data = JSON.parse(json);
      for (const row of data) {
        map.set(row[0], row);
      }
    });
  }

  return fixtures;
}
