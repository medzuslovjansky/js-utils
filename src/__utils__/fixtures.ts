type FixtureGetter = () => [string, string, string, string][];
type StringsGetter = () => string[];

export const adjectives: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/adjectives.json');
export const nouns_feminine: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/nouns-feminine.json');
export const nouns_masculine_animate: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/nouns-masculine-animate.json');
export const nouns_masculine: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/nouns-masculine.json');
export const nouns_misc: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/nouns-misc.json');
export const nouns_neuter: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/nouns-neuter.json');
export const verbs_imperfect: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/verbs-imperfect.json');
export const verbs_perfect: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/verbs-perfect.json');
export const verbs_misc: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/verbs-misc.json');
export const pronouns_demonstrative: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-demonstrative.json');
export const pronouns_indefinite: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-indefinite.json');
export const pronouns_interrogative: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-interrogative.json');
export const pronouns_personal: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-personal.json');
export const pronouns_possessive: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-possessive.json');
export const pronouns_reciprocal: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-reciprocal.json');
export const pronouns_reflexive: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-reflexive.json');
export const pronouns_relative: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/pronouns-relative.json');
export const other: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/other.json');
export const word_forms: StringsGetter = () =>
  jest.requireActual('../__fixtures__/word-forms.json');
