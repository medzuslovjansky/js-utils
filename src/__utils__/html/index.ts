export * as adjective from './adjective';
export * as noun from './noun';
export * as verb from './verb';
