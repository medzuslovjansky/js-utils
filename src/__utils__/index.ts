export * from './toHTML';
export * from './findLemmaById';
export * as fixtures from './fixtures';
