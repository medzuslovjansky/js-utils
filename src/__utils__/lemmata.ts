import fs from 'node:fs';
import path from 'node:path';
import { parsePos } from '../partOfSpeech';

const fixturesDir = path.join(__dirname, '../../src/__fixtures__');

let fixtures: Map<string, string[]> | undefined;

export function findLemmaById(id: string): string {
  return initFixtures().get(id)?.[2] ?? id;
}

export function findPosTags(id: string): string[] {
  const pos = initFixtures().get(id)?.[1];
  if (!pos) return [];

  const { name, ...attributes } = parsePos(pos);
  const trueAttributes = Object.entries(attributes)
    .filter(([_, value]) => value)
    .map(([key]) => key);

  return [name, ...trueAttributes];
}

function initFixtures() {
  if (!fixtures) {
    const map = (fixtures = new Map());
    fs.readdirSync(fixturesDir).forEach((name) => {
      const jsonPath = path.join(fixturesDir, name);
      const json = fs.readFileSync(jsonPath, 'utf-8');
      const data = JSON.parse(json);
      for (const row of data) {
        map.set(row[0], row);
      }
    });
  }

  return fixtures;
}
