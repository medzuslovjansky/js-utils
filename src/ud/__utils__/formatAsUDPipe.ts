import _ from 'lodash';
import { Word } from '../types';

export function formatAsUDPipe(data: Word[]) {
  const headers = ['Form', 'Lemma', 'uPosTag', 'Feats'];
  const rows = data.map((item) => [
    item.form,
    item.lemma,
    item.uPosTag,
    _.chain(item.feats)
      .toPairs()
      .map((pair) => _.upperFirst(pair.join('=')))
      .join('|')
      .value(),
  ]);

  const headerRow = `${headers.join('\t')}`;
  const dataRows = rows.map((row) => row.join('\t')).join('\n');

  return `${headerRow}\n${dataRows}`;
}
