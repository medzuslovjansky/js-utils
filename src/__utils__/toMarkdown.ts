import vm from 'node:vm';
import type { TestCaseResult } from '@jest/reporters';
import { transliterate } from '..';
import { toMarkdown as verb } from '../verb/__utils__/toMarkdown';
import { SteenVerbParadigm } from '../verb';

type Tuple<T> = [T, T];
type TupleProcessor<T> = (diff: Tuple<T>) => string;

export function toMarkdown(test: TestCaseResult) {
  const processor = getMarkdownProcessor(test);
  const tables = processor ? getBeforeAndAfter(test).map(processor) : [];
  return transliterate(
    tables.map((table) => table + '\n\n').join(''),
    'art-Latn-x-interslv-etym',
  );
}

function getBeforeAndAfter(test: TestCaseResult): Tuple<unknown>[] {
  return (test.failureDetails as any[])
    .map((failure) => failure?.matcherResult)
    .filter(
      (matcherResult) =>
        matcherResult && matcherResult.expected && matcherResult.actual,
    )
    .map((failure) =>
      vm.runInNewContext(`[${failure.expected}, ${failure.actual}]`),
    );
}

function getMarkdownProcessor(
  test: TestCaseResult,
): TupleProcessor<unknown> | undefined {
  const partOfSpeech = test.ancestorTitles[0];
  switch (partOfSpeech) {
    case 'verb':
      return (tuple) => verb(tuple as Tuple<SteenVerbParadigm>);
    default:
      return;
  }
}
