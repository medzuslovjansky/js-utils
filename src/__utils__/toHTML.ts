import vm from 'node:vm';

import type { TestCaseResult } from '@jest/reporters';

import { SteenVerbParadigm } from '../verb';
import { SteenNounParadigm } from '../noun';
import { SteenAdjectiveParadigm } from '../adjective';

import * as html from './html';

type Tuple<T> = [T, T];
type TupleProcessor<T> = (diff: Tuple<T>) => string;

export function toHTML(test: TestCaseResult) {
  const renderer = getRenderer(test);
  const figures = renderer ? getBeforeAndAfter(test).map(renderer) : [];
  return figures.map((figure) => figure + '\n\n').join('');
}

function getBeforeAndAfter(test: TestCaseResult): Tuple<unknown>[] {
  return (test.failureDetails as any[])
    .map((failure) => failure?.matcherResult)
    .filter(
      (matcherResult) =>
        matcherResult && matcherResult.expected && matcherResult.actual,
    )
    .map((failure) =>
      vm.runInNewContext(`[${failure.expected}, ${failure.actual}]`),
    );
}

function getRenderer(
  test: TestCaseResult,
): TupleProcessor<unknown> | undefined {
  const partOfSpeech = test.ancestorTitles[0];
  switch (partOfSpeech) {
    case 'verb':
      return (tuple) =>
        html.verb.renderDiff(...(tuple as Tuple<SteenVerbParadigm>));
    case 'noun':
      return (tuple) =>
        html.noun.renderDiff(...(tuple as Tuple<SteenNounParadigm>));
    case 'adjective':
      return (tuple) =>
        html.adjective.renderDiff(...(tuple as Tuple<SteenAdjectiveParadigm>));
    default:
      return;
  }
}
