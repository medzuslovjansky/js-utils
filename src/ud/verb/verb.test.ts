import { forms } from '.';
import { formatAsUDPipe } from '../__utils__';

describe('ud/verb', () => {
  test('forms', () => {
    const result = forms('praviti', '', 'v.tr. ipf.');
    expect(formatAsUDPipe(result)).toMatchSnapshot();
  });
});
