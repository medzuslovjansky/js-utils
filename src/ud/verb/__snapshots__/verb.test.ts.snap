// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ud/verb forms 1`] = `
"Form	Lemma	uPosTag	Feats
pravitì	pravitì	VERB	Aspect=Imp|VerbForm=Inf
pravjų, pravim	pravitì	VERB	Aspect=Imp|Mood=Ind|Tense=Pres|Person=1|Number=Sing
praviš	pravitì	VERB	Aspect=Imp|Mood=Ind|Tense=Pres|Person=2|Number=Sing
pravi	pravitì	VERB	Aspect=Imp|Mood=Ind|Tense=Pres|Person=3|Number=Sing
pravimȯ	pravitì	VERB	Aspect=Imp|Mood=Ind|Tense=Pres|Person=1|Number=Plur
pravite	pravitì	VERB	Aspect=Imp|Mood=Ind|Tense=Pres|Person=2|Number=Plur
pravęt	pravitì	VERB	Aspect=Imp|Mood=Ind|Tense=Pres|Person=3|Number=Plur
pravi	pravitì	VERB	Aspect=Imp|Mood=Imp|Person=2|Number=Sing
pravimȯ	pravitì	VERB	Aspect=Imp|Mood=Imp|Person=1|Number=Plur
pravite	pravitì	VERB	Aspect=Imp|Mood=Imp|Person=2|Number=Plur"
`;
